### PR TITLE
feat: add deleted to table Doc

### LINKED ISSUE #8: Add `deleted` field to doc schema
See https://github.com/digidem/mapeo-core-next/issues/137

### PR DESCRIPTION
**Pull request by tomasciccola**
would close #8

### CODE REVIEW

**Comment by tomasciccola** [comment]
I intuitively added `deleted` to the `IndexableDocument` type on `index.js`, but tests started to fail. 
On `docSchemas` I also added `deleted` as a `notnull:0` field with a type of `INTEGER` (since sqlite stores booleans as ints `1 | 0`), but there's probably something else that I'm missing.....
@gmaclennan any pointers?

**Comment by tomasciccola** [comment]
> CI tests are failing because of the change to package-lock

`npm ERR! Cannot read property '@types/better-sqlite3' of undefined`
is that what this error refers to??

> The approach sounds right, but I don't think the code you have done is pushed? 

Is there any more to add besides the `IndexableDocument` type? 

> Tests will need updated of course to include tests for deleted docs. 

For starters,  if we add `deleted` as an optional, wouldn't all test pass?

File: index.js [hunk 1]
```diff
@@ -315,14 +317,16 @@ function assertValidSchema(db, { docTableName, backlinkTableName }) {
...
+    if (info.notnull) {
+      assert(column, `Table '${tableName}' must have a column '${name}'`)
+      for (const [prop, value] of Object.entries(info)) {
+        assert(
+          // @ts-ignore
+          column[prop] === value,
+          // @ts-ignore
+          `Column '${name}' in table '${tableName}' should have ${prop}=${value}, but instead ${prop}=${column[prop]}`
+        )
+      }
```
Comment (gmaclennan) [inline index.js:329]: Not sure why you are doing this? We should check that all the columns exist in the table.

File: index.js [hunk 2]
```diff
@@ -28,6 +29,7 @@ const docSchema = {
   links: { type: 'TEXT', notnull: 1, dflt_value: null, pk: 0 },
   forks: { type: 'TEXT', notnull: 1, dflt_value: null, pk: 0 },
   updatedAt: { type: 'TEXT', notnull: 1, pk: 0 },
+  deleted: { type: 'INTEGER', notnull: 0, pk: 0 },
```
Comment (gmaclennan) [inline index.js:32]: Not sure if it makes sense to make it notnull? Either way you will have to decide what to do with the `null` value returned on read.

File: index.js (hunk 1 above)
Comment (tomasciccola) [reply index.js:329]: oh, I thought that since `deleted` was optional, then it shouldn't be created directly with the table

File: index.js (hunk 2 above)
Comment (tomasciccola) [reply index.js:32]: Ok, so despite `deleted` being optional and always `false` in the schema. In the table is required and initialized to false, right?

**Review by gmaclennan — CHANGES_REQUESTED** [review]
This should return deleted as a boolean - tests are expecting a number `0`.

**Review by gmaclennan — CHANGES_REQUESTED** [review]
Can you include a test that actually tests when `deleted=true`? Maybe in the scenarios?

**Comment by tomasciccola** [comment]
I made `deleted` `notnull:1` and a required field on `IndexableDocument`
I also updated the tests to contemplate that. 
The only doubt that I have is that when writting a Doc with `deleted: false`, when we read it (with `.getDoc()`) we get `deleted: 0`, which is the desired behaviour since Sqlite treats booleans as ints (0 | 1). 
But, should we cast the returned value to a boolean when we read it back or should we leave it as that? 
Also, what tests should we add regarding `deleted`? I could modify the tests so we sometime test `deleted: true` values (expecting `deleted: 1` when read back), but other than that I dunno...

**Review by gmaclennan — APPROVED** [review]
Great stuff 👍

**Comment by gmaclennan** [comment]
> I made `deleted` `notnull:1` and a required field on `IndexableDocument`
> I also updated the tests to contemplate that.
> The only doubt that I have is that when writting a Doc with `deleted: false`, when we read it (with `.getDoc()`) we get `deleted: 0`, which is the desired behaviour since Sqlite treats booleans as ints (0 | 1).
> But, should we cast the returned value to a boolean when we read it back or should we leave it as that?
> Also, what tests should we add regarding `deleted`? I could modify the tests so we sometime test `deleted: true` values (expecting `deleted: 1` when read back), but other than that I dunno...

@tomasciccola just wanted to follow up on this so that there is a record of it. Because we read this data via drizzle, that takes care of converting 1 | 0 into a boolean on read. It's the same with the JSON fields (drizzle does JSON parse).

**Pull request merged**